File: spark-in-action/java/src/main/java/transform/document/FlattenJsonApp.java
package transform.document;

import static org.apache.spark.sql.functions.explode;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

public class FlattenJsonApp {
    public static final String ARRAY_TYPE = "Array";
    public static final String STRUCT_TYPE = "Struc";

    public static void main(String[] args) {
        FlattenJsonApp app = new FlattenJsonApp();
        app.start();
    }

    private void start() {
        // Creates a session on a local master
        SparkSession spark = SparkSession.builder()
                .appName("Flatten JSON data")
                .master("local")
                .getOrCreate();

        spark.sparkContext().setLogLevel("ERROR");

        // Reads a JSON, stores it in a dataframe
        Dataset<Row> invoicesDf = spark.read()
                .format("json")
                .option("multiline", "true")
                .load("data/ch13/json/nested_array.json");

        invoicesDf.show(3);
        invoicesDf.printSchema();

        Dataset<Row> flatInvoicesDf = flattenNestedStructure(spark, invoicesDf);
        flatInvoicesDf.show(20, false);
        flatInvoicesDf.printSchema();
    }

    public static Dataset<Row> flattenNestedStructure(
            SparkSession spark,
            Dataset<Row> df) {
        boolean recursion = false;

        Dataset<Row> processedDf = df;
        StructType schema = df.schema();
        StructField fields[] = schema.fields();
        for (StructField field : fields) {
            switch (field.dataType().toString().substring(0, 5)) {
                case ARRAY_TYPE:
                    processedDf = processedDf
                            .withColumnRenamed(field.name(), field.name() + "_tmp");
                    processedDf = processedDf.withColumn(
                            field.name(),
                            explode(processedDf.col(field.name() + "_tmp")));
                    processedDf = processedDf.drop(field.name() + "_tmp");
                    recursion = true;
                    break;
                case STRUCT_TYPE:
                    // Mapping
                    StructField[] structFields = ((StructType) field.dataType()).fields();
                    for (StructField structField : structFields) {
                        processedDf = processedDf.withColumn(
                                field.name() + "_" + structField.name(),
                                processedDf.col(field.name() + "." + structField.name()));
                    }
                    processedDf = processedDf.drop(field.name());
                    recursion = true;
                    break;
            }
        }
        if (recursion) {
            processedDf = flattenNestedStructure(spark, processedDf);
        }
        return processedDf;
    }
}

File: spark-in-action/scala/src/main/scala/transform/document/FlattenJsonApp.scala
package transform.document

import org.apache.spark.sql.{SparkSession, DataFrame, functions => F}

object FlattenJsonApp {
  def main(args: Array[String]): Unit = {
    val spark: SparkSession = SparkSession.builder
      .appName("Display Flattened JSON data")
      .master("local")
      .getOrCreate

    spark.sparkContext.setLogLevel("ERROR")

    val df = spark.read
      .format("json")
      .option("multiline", "true")
      .load("data/ch13/json/shipment.json")

    val df2 = df
      .withColumn("supplier_name", F.col("supplier.name"))
      .withColumn("supplier_city", F.col("supplier.city"))
      .withColumn("supplier_state", F.col("supplier.state"))
      .withColumn("supplier_country", F.col("supplier.country"))
      .drop("supplier")
      .withColumn("customer_name", F.col("customer.name"))
      .withColumn("customer_city", F.col("customer.city"))
      .withColumn("customer_state", F.col("customer.state"))
      .withColumn("customer_country", F.col("customer.country"))
      .drop("customer")
      .withColumn("items", F.explode(F.col("books")))

    val df3 = df2
      .withColumn("qty", F.col("items.qty"))
      .withColumn("title", F.col("items.title"))
      .drop("items", "books")

    df3.show(5, false)
    df3.printSchema()

    df3.createOrReplaceTempView("shipment_detail")

    val sqlQuery =
      """
        |SELECT COUNT(*) AS bookCount
        |FROM shipment_detail
        |""".stripMargin
    val bookCountDf = spark.sql(sqlQuery)

    bookCountDf.show

    spark.stop
  }

}
